sdl_android/src/main/java/com/smartdevicelink/api/screen/TextAndGraphicManager.java

```diff
@@ -0,0 +1,893 @@
+			public void onError(String info) {
+				Log.w(TAG, "DISPLAY Capability cannot be retrieved:");
+				displayCapabilities = null;
+			}
+		};
+		this.internalInterface.addOnSystemCapabilityListener(SystemCapabilityType.DISPLAY, onDisplayCapabilitiesListener);
+	}
+
+	// Upload / Send
+
+	protected void update(CompletionListener listener) {
+
+		// check if is batch update
+		if (batchingUpdates) {
+			return;
+		}
+
+		if (isDirty){
+			isDirty = false;
+			sdl_update(listener);
+		}
+	}
+	
+	private void sdl_update(CompletionListener listener){
+
+		// make sure hmi is not none
+		if (currentHMILevel == null || currentHMILevel == HMILevel.HMI_NONE){
+			Log.v(TAG, "Trying to send show on HMI_NONE, waiting for full");
+			pendingHMIFull = true;
+			if (listener != null){
+				pendingHMIListener = listener;
+			}
+			return;
+		}
+
+		Log.v(TAG, "Updating Text and Graphics");
+		if (inProgressUpdate != null){
+
+			Log.v(TAG, "In progress update exists, queueing update");
+			if (queuedUpdateListener != null){
+
+				Log.v(TAG, "Queued update already exists, superseding previous queued update");
+				queuedUpdateListener.onComplete(false);
+				queuedUpdateListener = null;
+			}
+
+			if (listener != null){
+				queuedUpdateListener = listener;
+			}else{
+				hasQueuedUpdate = true;
+			}
+			return;
+		}
+
+		Show fullShow = new Show();
+		fullShow.setAlignment(textAlignment);
+		fullShow = assembleShowText(fullShow);
+		fullShow = assembleShowImages(fullShow);
+
+		inProgressListener = listener;
+
+		if (!shouldUpdatePrimaryImage() && !shouldUpdateSecondaryImage()){
+
+			Log.v(TAG, "No Images to send, only sending text");
+			inProgressUpdate = extractTextFromShow(fullShow);
+			sendShow();
+
+		}else if (isArtworkUploadedOrDoesntExist(primaryGraphic) && ( secondaryGraphic == blankArtwork || isArtworkUploadedOrDoesntExist(secondaryGraphic))){
+
+			Log.v(TAG, "Images already uploaded, sending full update");
+			// The files to be updated are already uploaded, send the full show immediately
+			inProgressUpdate = fullShow;
+			sendShow();
+		} else{
+
+			Log.v(TAG, "Images need to be uploaded, sending text and uploading images");
+			// We need to upload or queue the upload of the images
+			// start uploading images
+			inProgressUpdate = fullShow;
+			final Show thisUpdate = fullShow;
+
+			uploadImages(new CompletionListener() {
+				@Override
+				public void onComplete(boolean success) {
+					if (!success){
+						Log.e(TAG, "Error uploading image");
+						inProgressUpdate = extractTextFromShow(inProgressUpdate);
+						sendShow();
+					}
+					// Check if queued image update still matches our images (there could have been a new Show in the meantime)
+					// and send a new update if it does. Since the images will already be on the head unit, the whole show will be sent
+					if ((thisUpdate.getGraphic() != null && queuedImageUpdate.getGraphic() != null) && thisUpdate.getGraphic().equals(queuedImageUpdate.getGraphic()) ||
```

Comment:
No need for second null check. Logic check can be as followed: 
`thisUpdate.getGraphic() != null && thisUpdate.getGraphic().equals(queuedImageUpdate.getGraphic())`